File: expansion_board/spi/flash/python/w25qxx.py
###############################################
#
#  file: w25qxx.py
#  update: 2024-11-14
#  usage: 
#      sudo python w25qxx.py
#
###############################################

import time
import spidev
import gpiod

W25QXX_80_ID = 0xEF13
W25QXX_16_ID = 0xEF14
W25QXX_32_ID = 0xEF15
W25QXX_64_ID = 0xEF16
W25QXX_128_ID = 0xEF17

W25QXX_PAGE_SIZE = 256
W25QXX_SECTOR_SIZE = 4096

W25QXX_WRITE_ENABLE = 0x06
W25QXX_SECTOR_ERASE_4KB = 0x20
W25QXX_PAGE_PROGRAM = 0x02
W25QXX_READ_DATA = 0x03
W25QXX_READ_STATUS_REGISTER_1 = 0x05

# W25QXX类
class W25QXX:
    def __init__(self, spi_bus, cs_pin, cs_chip):
        self.spi_bus = spi_bus
        self.cs_pin = cs_pin
        self.cs_chip = cs_chip
        self.gpiochip = gpiod.Chip(cs_chip, gpiod.Chip.OPEN_BY_NUMBER)                              # 选择正确的GPIO芯片
        self.cs_line = self.gpiochip.get_line(self.cs_pin)                                          # 设置CS引脚
        self.cs_line.request(consumer="w25qxx_spi", type=gpiod.LINE_REQ_DIR_OUT, default_vals=[1])  # 拉高CS，默认不选择

        self.spi = spidev.SpiDev()          # 创建SPI对象
        self.spi.open(spi_bus, 0)           # 设置SPI总线

        self.spi.max_speed_hz = 1000000     # 设置SPI速度
        self.spi.mode = 0b00                # SPI模式0

        self.device_id = None

    def w25qxx_init(self):
        """初始化W25QXX并获取设备ID"""
        device_id = self.w25qxx_read_id()
        if device_id in [W25QXX_80_ID, W25QXX_16_ID, W25QXX_32_ID, W25QXX_64_ID, W25QXX_128_ID]:
            self.device_id = device_id
            print(f"Device ID: {hex(device_id)}")
            return 0
        else:
            print("Unknown device ID")
            return -1

    def w25qxx_read_id(self):
        """读取设备ID"""
        send_buf = [0x90, 0x00, 0x00, 0x00]         # 发送命令
        self.cs_line.set_value(0)                   # 拉低CS，开始SPI通信
        self.spi.xfer(send_buf)                     # 发送数据
        receive_data = self.spi.readbytes(2)        # 读取2字节数据
        self.cs_line.set_value(1)                   # 拉高CS，结束SPI通信
        device_id = (receive_data[0] << 8) | receive_data[1]
        return device_id

    def w25qxx_wait_state_free(self):
        """等待W25QXX状态寄存器的忙标志位清除"""
        while True:
            self.cs_line.set_value(0)                       # 拉低CS，开始SPI通信
            self.spi.xfer([W25QXX_READ_STATUS_REGISTER_1])  # 发送命令
            recv_buf = self.spi.readbytes(1)                # 读状态
            self.cs_line.set_value(1)                       # 拉高CS，结束SPI通信
            if recv_buf[0] & 0x01 == 0:                     # 判断忙标志位
                break
            time.sleep(0.01)

    def w25qxx_write_enable(self):
        """使能W25QXX写操作"""
        self.cs_line.set_value(0)
        self.spi.xfer([W25QXX_WRITE_ENABLE])                # 发送写使能命令
        self.cs_line.set_value(1)

    def w25qxx_sector_erase(self, addr):
        """对指定地址进行4KB扇区擦除"""
        self.w25qxx_write_enable()
        send_buf = [W25QXX_SECTOR_ERASE_4KB, addr >> 16, addr >> 8, addr]
        self.cs_line.set_value(0)
        self.spi.xfer(send_buf)                             # 发送扇区擦除命令
        self.cs_line.set_value(1)
        self.w25qxx_wait_state_free()

    def w25qxx_page_write(self, addr, write_buff):
        """向指定页写入数据"""
        if len(write_buff) < 1 or len(write_buff) > W25QXX_PAGE_SIZE:
            print("Page write error, invalid data length. len = ", len(write_buff))
            return
        self.w25qxx_write_enable()
        send_buf = [W25QXX_PAGE_PROGRAM, addr >> 16, addr >> 8, addr]
        send_buf.extend(write_buff)
        self.cs_line.set_value(0)
        self.spi.xfer(send_buf)
        self.cs_line.set_value(1)
        self.w25qxx_wait_state_free()

    def w25qxx_npage_write(self, addr, write_buff, write_len):
        """执行多页写入操作，考虑到可能的跨页和跨扇区写入"""
        page_size = W25QXX_PAGE_SIZE            # 每页的字节数
        sector_size = W25QXX_SECTOR_SIZE        # 每个扇区的字节数

        offset = 0                              # 当前写入的偏移量
        total_len = write_len                   # 总写入长度

        sector = -1                             # 上次写入的扇区编号
        current_sector = 0                      # 当前要写入的扇区编号

        if write_len < 1:
            print("Page write error, data length is invalid!")
            return

        while total_len > 0:
            # 计算当前数据的目标扇区编号
            current_sector = (addr + offset) // sector_size

            if sector != current_sector:
                sector = current_sector
                # print(f"Erasing sector {sector * sector_size}...")
                self.w25qxx_sector_erase(sector * sector_size)  # 擦除目标扇区

            # 每次最多写入 page_size 字节，剩余数据长度不超过 page_size 时，调整写入长度
            write_bytes = min(page_size, total_len)

            # 调用 page_write 进行单页写入
            # print(f"Writing data to address {hex(addr + offset)}: {write_buff[offset:offset + write_len]}")
            self.w25qxx_page_write(addr + offset, write_buff[offset:offset + write_bytes])

            # 更新偏移量和剩余数据长度
            offset += write_bytes
            total_len -= write_bytes

    def w25qxx_read_byte_data(self, addr, read_len):
        """从指定地址读取数据"""
        send_buf = [W25QXX_READ_DATA, addr >> 16, addr >> 8, addr]
        self.cs_line.set_value(0)  
        self.spi.xfer(send_buf)  
        receive_data = self.spi.readbytes(read_len)  
        self.cs_line.set_value(1)  
        return receive_data

File: expansion_board/experinment/experiment_4_RGB/spi/c/driver/w25qxx.c
/*
*
*   file: w25qxx.c
*   updata: 2024-11-11
*
*/

#include "w25qxx.h"

struct gpiod_chip *cs_gpiochip;        
struct gpiod_line *cs_gpioline;

spi_operations_t *w25qxx_spi_ops = NULL;

/*****************************
 * @brief : 初始化 W25QXX 芯片的 CS 引脚
 * @param : 无
 * @return: 成功返回0 失败返回-1
 *****************************/
static int w25qxx_cs_pin_init(const char *cs_chip, unsigned int cs_pin)
{
    int ret; 

    /* cs pin init */
    cs_gpiochip = gpiod_chip_open(cs_chip);
    if(cs_gpiochip == NULL)
    {
        printf("gpiod_chip_open error\n");
        return -1;
    }
    cs_gpioline = gpiod_chip_get_line(cs_gpiochip, cs_pin);
    if(cs_gpioline == NULL)
    {
        printf("gpiod_chip_get_line error\n");
        return -1;
    }
    ret = gpiod_line_request_output(cs_gpioline, "cs_gpioline", 1);
    if(ret < 0)
    {
        printf("gpiod_line_request_output error : cs_gpioline\n");
        return -1;
    }

    return 0;
}

/*****************************
 * @brief : 读取 W25QXX 芯片的 ID
 * @param : 无
 * @return: 返回 W25QXX 芯片的设备 ID
 *****************************/
static unsigned int w25qxx_read_id()
{
    unsigned char send_buf[4] = {0x00, 0x00, 0x00, 0x00};   
    unsigned char recv_buf[2] = {0};
    unsigned int id;

    if(w25qxx_spi_ops == NULL)
        return -1;

    send_buf[0] = W25QXX_MANUFACTURER_DEVICE_ID;            // Manufacturer/Device ID

    gpiod_line_set_value(cs_gpioline, 0);

    pthread_mutex_lock(w25qxx_spi_ops->mutex);
    w25qxx_spi_ops->spi_write_then_read(send_buf, sizeof(send_buf), recv_buf, sizeof(recv_buf));
    pthread_mutex_unlock(w25qxx_spi_ops->mutex);

    gpiod_line_set_value(cs_gpioline, 1);

    id = (recv_buf[0] << 8) | recv_buf[1];

    return id;
}

/*****************************
 * @brief : 等待 W25QXX 状态寄存器的忙标志位清除
 * @param : 无
 * @return: 无返回值
 *****************************/
static void w25qxx_wait_state_free()
{
    unsigned char send_buf[1] = {W25QXX_READ_STATUS_REGISTER_1};   
    unsigned char recv_buf[1] = {0};

    if(w25qxx_spi_ops == NULL)
        return;

    gpiod_line_set_value(cs_gpioline, 0);

    pthread_mutex_lock(w25qxx_spi_ops->mutex);
    do
    {
        w25qxx_spi_ops->spi_write_then_read(send_buf, sizeof(send_buf), recv_buf, sizeof(recv_buf));
    } while (recv_buf[0] & 0x01);
    pthread_mutex_unlock(w25qxx_spi_ops->mutex);

    gpiod_line_set_value(cs_gpioline, 1);
}

/*****************************
 * @brief : 使能 W25QXX 的写操作
 * @param : 无
 * @return: 无返回值
 *****************************/
static void w25qxx_write_enable()
{
    if(w25qxx_spi_ops == NULL)
        return;

    gpiod_line_set_value(cs_gpioline, 0);

    pthread_mutex_lock(w25qxx_spi_ops->mutex);
    w25qxx_spi_ops->spi_write_byte_data(W25QXX_WRITE_ENABLE);
    pthread_mutex_unlock(w25qxx_spi_ops->mutex);

    gpiod_line_set_value(cs_gpioline, 1);
}

/*****************************
 * @brief : 对指定地址所在的扇区执行 4KB 擦除
 * @param : addr - 扇区的起始地址
 * @return: 无返回值
 *****************************/
void w25qxx_sector_erase(unsigned int addr)
{
    if(w25qxx_spi_ops == NULL)
        return;

    w25qxx_write_enable();

    gpiod_line_set_value(cs_gpioline, 0);

    pthread_mutex_lock(w25qxx_spi_ops->mutex);
    w25qxx_spi_ops->spi_write_byte_data(W25QXX_SECTOR_ERASE_4KB);   // 4K扇区擦除指令
    w25qxx_spi_ops->spi_write_byte_data(addr>>16);                  // 24~16地址
	w25qxx_spi_ops->spi_write_byte_data(addr>>8);                   // 16~8地址
	w25qxx_spi_ops->spi_write_byte_data(addr);                      // 8~0地址
    pthread_mutex_unlock(w25qxx_spi_ops->mutex);

    gpiod_line_set_value(cs_gpioline, 1);

    w25qxx_wait_state_free();                       // 等待操作完成
}

/*****************************
 * @brief : 向指定页写入数据
 * @param : addr - 写入数据的起始地址
 *          write_buff - 待写入的数据缓冲区
 *          len - 待写入数据的长度，最大为 256 字节
 * @return: 无返回值
 * @note  : 向指定页写入数据，若数据超过 256 字节需使用 w25qxx_npage_write() 函数。
 *****************************/
void w25qxx_page_write(unsigned int addr, unsigned char *write_buff, int len)
{
    int i;

    if(w25qxx_spi_ops == NULL)
        return;

    if(len < 1 || len > W25QXX_PAGE_SIZE)
    {
        printf("page write error, data length is invalid!\n");
        return;
    }

    w25qxx_write_enable();

    gpiod_line_set_value(cs_gpioline, 0);

    pthread_mutex_lock(w25qxx_spi_ops->mutex);
    w25qxx_spi_ops->spi_write_byte_data(W25QXX_PAGE_PROGRAM);       // 页编程指令
    w25qxx_spi_ops->spi_write_byte_data(addr>>16);                  // 24~16地址
	w25qxx_spi_ops->spi_write_byte_data(addr>>8);                   // 16~8地址
	w25qxx_spi_ops->spi_write_byte_data(addr);                      // 8~0地址

    for(i = 0; i < len; i++)
	{
		w25qxx_spi_ops->spi_write_byte_data(write_buff[i]);         // 写数据	
	}
	pthread_mutex_unlock(w25qxx_spi_ops->mutex);

    gpiod_line_set_value(cs_gpioline, 1);

    w25qxx_wait_state_free();                       // 等待操作完成
}

/*****************************
 * @brief : 执行多页写入操作，自动处理扇区擦除和分页写入
 * @param : addr - 写入数据的起始地址
 *          write_buff - 存放待写入数据的缓冲区
 *          len - 待写入数据的长度，单位：字节
 * @return: 无返回值
 * @note  : 该函数会根据地址自动处理每次写入的页和扇区的擦除操作。
 *           - 每次写入数据时，最大支持 256 字节（页大小），若数据超出当前页则会继续写入到下一页
 *           - 写入前会判断是否需要擦除当前扇区
 *           - 支持跨页写入，当数据超出当前页时，会继续写入下一个页
 *           - 擦除的最小单位为4KB（一个扇区）
 *           - 写入过程中，如果目标地址跨越多个扇区，会自动进行扇区擦除
 *****************************/
void w25qxx_npage_write(unsigned int addr, unsigned char *write_buff, int len)
{
    int page_size = W25QXX_PAGE_SIZE;               // 每页最大写入256字节
    int offset = 0;                                 // 当前写入数据的偏移量
    int write_len;                                  // 每次写入的数据长度

    int sector = -1;                                // 上次写入的扇区编号
    int current_sector;                             // 当前要写入的扇区编号

    if(w25qxx_spi_ops == NULL)
        return;

    if (len < 1) {
        printf("page write error, data length is invalid!\n");
        return;
    }

    while (len > 0) 
    {
        current_sector = (addr + offset) / W25QXX_SECTOR_SIZE;      // 计算当前数据的目标扇区编号
        if (sector != current_sector)                               // 判断是否跨越扇区，需擦除新的扇区
        {
            sector = current_sector;                
            w25qxx_sector_erase(sector * W25QXX_SECTOR_SIZE);       // 擦除目标扇区
        }
        
        write_len = (len > page_size) ? page_size : len;            // 每次最多写 256 字节，剩余数据长度不超过 256 字节时，调整写入长度

        w25qxx_page_write(addr + offset, write_buff + offset, write_len);

        offset += write_len;                        // 更新偏移量
        len -= write_len;                           // 更新剩余数据长度
    }
}

/*****************************
 * @brief : 从指定地址读取数据
 * @param : addr - 读取数据的起始地址
 *          read_buff - 用于存储读取数据的缓冲区
 *          len - 读取的数据长度
 * @return: 无返回值
 *****************************/
void w25qxx_read_byte_data(unsigned int addr, unsigned char *read_buff, int len)
{
    unsigned char send_buf[4];   

    if(w25qxx_spi_ops == NULL)
        return;

    send_buf[0] = W25QXX_READ_DATA;
    send_buf[1] = addr >> 16;
    send_buf[2] = addr >> 8;
    send_buf[3] = addr;

    gpiod_line_set_value(cs_gpioline, 0);

    pthread_mutex_lock(w25qxx_spi_ops->mutex);
    w25qxx_spi_ops->spi_write_then_read(send_buf, sizeof(send_buf), read_buff, len);
    pthread_mutex_unlock(w25qxx_spi_ops->mutex);

    gpiod_line_set_value(cs_gpioline, 1);
}

/*****************************
 * @brief : 初始化 W25QXX 芯片
 * @param : cs_chip - GPIO 控制器名称
 *          cs_pin - 片选引脚号
 * @return: 成功返回 0，失败返回 -1
 *****************************/
int w25qxx_init(const char *cs_chip, unsigned int cs_pin)
{
    int ret;
    unsigned int device_id;

    ret = w25qxx_cs_pin_init(cs_chip, cs_pin);
    if(ret < 0)
        return -1;

    device_id = w25qxx_read_id();
    switch (device_id)
    {
        case W25QXX_80_ID:
            printf("Find the W25Q80");
            break;
        case W25QXX_16_ID:
            printf("Find the W25Q16");
            break;
        case W25QXX_32_ID:
            printf("Find the W25Q32");
            break;
        case W25QXX_64_ID:
            printf("Find the W25Q64");
            break;
        case W25QXX_128_ID:
            printf("Find the W25Q128");
            break;
        default:
            printf("No storage device found!\n");
            return -1;
    }
    printf(" , id : 0x%x\n", device_id);

    return 0;
}

/*****************************
 * @brief : 注册 SPI 操作函数
 * @param : ops - SPI 操作函数结构体的指针，用于设置 W25QXX 的 SPI 通信接口
 * @return: 无返回值
 *****************************/
void w25qxx_register_spi_operations(spi_operations_t *ops)
{
    w25qxx_spi_ops = ops;
}